File: misc/leetcode/next-permutation.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        k = None
        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j] and \
                        (k is None or nums[j] < nums[k]):
                    k = j
            if k is not None:
                (nums[i], nums[k]) = (nums[k], nums[i])
                nums[i + 1:] = sorted(nums[i + 1:])
                break
        if k is None: nums.sort()
        # return nums


if __name__ == '__main__':
    s = Solution()
    print(s.nextPermutation([1, 2, 3]))
    print(s.nextPermutation([3, 2, 1]))
    print(s.nextPermutation([1, 1, 5]))
    print(s.nextPermutation([1, 1, 5, 4, 0, 4]))

File: leetcode/cpp/next-permutation.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <vector>
using namespace std;

class Solution {
   public:
    void nextPermutation(vector<int>& nums) {
        if (nums.size() == 0) return;
        // find swap position.
        int i = nums.size() - 1;
        for (; i >= 1; i--) {
            if (nums[i] > nums[i - 1]) {
                break;
            }
        }
        if (i == 0) {
            sort(nums.begin(), nums.end());
            return;
        }
        // find another swap position.
        int min_value = nums[i];
        int min_index = i;
        for (int j = i + 1; j < nums.size(); j++) {
            if (nums[j] <= nums[i - 1]) continue;
            if (nums[j] < min_value) {
                min_value = nums[j];
                min_index = j;
            }
        }
        swap(nums[min_index], nums[i - 1]);
        sort(nums.begin() + i, nums.end());
    }
};
